packages/core/src/browser/tree/tree-widget.tsx

```diff
@@ -1135,15 +1146,19 @@ export class TreeWidget extends ReactWidget implements StatefulWidget {
                         this.model.selectNode(node);
                     }
                 }
-                if (this.isExpandable(node) && !shiftMask && !ctrlCmdMask) {
-                    this.model.toggleNodeExpansion(node);
+                if (!this.props.expandOnlyOnExpansionToggleClick) {
+                    if (this.isExpandable(node) && !shiftMask && !ctrlCmdMask) {
+                        this.model.toggleNodeExpansion(node);
+                    }
                 }
             } else {
                 if (SelectableTreeNode.is(node)) {
                     this.model.selectNode(node);
                 }
-                if (this.isExpandable(node) && !this.hasCtrlCmdMask(event) && !this.hasShiftMask(event)) {
-                    this.model.toggleNodeExpansion(node);
+                if (!this.props.expandOnlyOnExpansionToggleClick) {
```

Comment:
This conditional expansion is repeated in both arms of the `props.multiSelect` conditional, so we might as well move it to after the `else`. Similarly, the shift and control masks are accessed in both arms (or outside the conditional, if you move this code out), so they could be defined above the conditional to expand their scope.